### PR TITLE
Fix propagation of build flags to unrelated targets

### PR DESCRIPTION
**Pull request by fmoc**
This PR makes sure flags set on our own targets won't be propagated to third-party targets which do not need them. In the present case, a new third-party dependency generated a `config.h` file, but instead of that one, ownCloud's was included, causing the build to fail.

I replaced all occurrences of `add_libraries`, `include_directories` etc. (`link_libraries` was not used anywhere).

### CODE REVIEW

File: shell_integration/dolphin/CMakeLists.txt [hunk 1]
```diff
@@ -43,6 +40,8 @@ add_library(${OWNCLOUDDOLPHINHELPER} SHARED ownclouddolphinpluginhelper.cpp)
 target_link_libraries(${OWNCLOUDDOLPHINHELPER} Qt5::Network Qt5::Gui)
 generate_export_header(${OWNCLOUDDOLPHINHELPER} BASE_NAME ownclouddolphinpluginhelper)
 install(TARGETS ${OWNCLOUDDOLPHINHELPER} LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR})
+target_include_directories(${OWNCLOUDDOLPHINHELPER} PUBLIC ${CMAKE_CURRENT_BINARY_DIR} ${PROJECT_BINARY_DIR} ${PROJECT_BINARY_DIR}/src)
```
Comment (TheOneRing) [inline shell_integration/dolphin/CMakeLists.txt:43]: https://cmake.org/cmake/help/latest/variable/CMAKE_INCLUDE_CURRENT_DIR.html
Will be needed in the dolphin CMakeLists.txt too as it has its own project() scope.
But should be covered by https://invent.kde.org/frameworks/extra-cmake-modules/-/blob/master/kde-modules/KDECMakeSettings.cmake#L194

File: shell_integration/dolphin/CMakeLists.txt [hunk 2]
```diff
@@ -43,6 +40,8 @@ add_library(${OWNCLOUDDOLPHINHELPER} SHARED ownclouddolphinpluginhelper.cpp)
...
+target_compile_definitions(${OWNCLOUDDOLPHINHELPER} PUBLIC -DQT_USE_FAST_CONCATENATION -DQT_USE_FAST_OPERATOR_PLUS)
```
Comment (TheOneRing) [inline shell_integration/dolphin/CMakeLists.txt:44]: Private

File: src/csync/CMakeLists.txt [hunk 3]
```diff
@@ -42,11 +33,49 @@ configure_file(csync_version.h.in ${CMAKE_CURRENT_BINARY_DIR}/csync_version.h)
...
+    PUBLIC -DQT_NO_URL_CAST_FROM_STRING
+    PUBLIC -DQT_NO_CAST_FROM_BYTEARRAY
+    PUBLIC -DQT_USE_QSTRINGBUILDER
+    PUBLIC -DQT_MESSAGELOGCONTEXT #enable function name and line number in debug output
+    PUBLIC -DQT_DEPRECATED_WARNINGS
+    PUBLIC -DQT_NO_FOREACH
+)
+
+if(MEM_NULL_TESTS)
+    target_compile_definitions(csync PUBLIC -DCSYNC_MEM_NULL_TESTS)
```
Comment (TheOneRing) [inline src/csync/CMakeLists.txt:56]: Can be removed, isn't referenced anywhere.

File: src/csync/CMakeLists.txt [hunk 4]
```diff
@@ -42,11 +33,49 @@ configure_file(csync_version.h.in ${CMAKE_CURRENT_BINARY_DIR}/csync_version.h)
...
+target_include_directories(csync
+    PUBLIC ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/std
+    # include root source dir in order to make src/common available to the targets
+    PUBLIC ${PROJECT_SOURCE_DIR}/src
+    # make sure auto-generated headers can be included
+    PUBLIC ${PROJECT_BINARY_DIR}
 )
 
+target_compile_definitions(csync
+    PUBLIC -DQT_NO_CAST_TO_ASCII
```
Comment (TheOneRing) [inline src/csync/CMakeLists.txt:45]: Private

File: src/csync/CMakeLists.txt [hunk 5]
```diff
@@ -42,11 +33,49 @@ configure_file(csync_version.h.in ${CMAKE_CURRENT_BINARY_DIR}/csync_version.h)
 
 add_library(csync SHARED ${common_SOURCES} ${csync_SRCS})
 
-target_include_directories(
-  csync
-  PUBLIC ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/std
+target_include_directories(csync
+    PUBLIC ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/std
```
Comment (TheOneRing) [inline src/csync/CMakeLists.txt:37]: Should be covered by https://invent.kde.org/frameworks/extra-cmake-modules/-/blob/master/kde-modules/KDECMakeSettings.cmake#L194
https://cmake.org/cmake/help/latest/variable/CMAKE_INCLUDE_CURRENT_DIR.html

File: src/csync/CMakeLists.txt [hunk 6]
```diff
@@ -42,11 +33,49 @@ configure_file(csync_version.h.in ${CMAKE_CURRENT_BINARY_DIR}/csync_version.h)
...
+    target_compile_definitions(csync PUBLIC -DCSYNC_MEM_NULL_TESTS)
+endif(MEM_NULL_TESTS)
+
+if(TOKEN_AUTH_ONLY)
+    message("Compiling with token authentication")
+    target_compile_definitions(csync PUBLIC -DTOKEN_AUTH_ONLY=1)
+endif()
+
+if(NO_MSG_HANDLER)
+    target_compile_definitions(csync PUBLIC -DNO_MSG_HANDLER=1)
```
Comment (TheOneRing) [inline src/csync/CMakeLists.txt:65]: Private in libsync where it is used.

File: src/csync/CMakeLists.txt [hunk 7]
```diff
@@ -42,11 +33,49 @@ configure_file(csync_version.h.in ${CMAKE_CURRENT_BINARY_DIR}/csync_version.h)
...
+    PUBLIC -DQT_NO_FOREACH
+)
+
+if(MEM_NULL_TESTS)
+    target_compile_definitions(csync PUBLIC -DCSYNC_MEM_NULL_TESTS)
+endif(MEM_NULL_TESTS)
+
+if(TOKEN_AUTH_ONLY)
+    message("Compiling with token authentication")
+    target_compile_definitions(csync PUBLIC -DTOKEN_AUTH_ONLY=1)
```
Comment (TheOneRing) [inline src/csync/CMakeLists.txt:61]: I hate this flag as the feature does not exist any more, this is an imaginary no ui config which probably won't even build.

File: src/libsync/CMakeLists.txt [hunk 8]
```diff
@@ -82,6 +77,12 @@ target_link_libraries(libsync PUBLIC
     csync
     Qt5::Core Qt5::Network
 )
+target_compile_definitions(libsync
+    PUBLIC -DQT_NO_CAST_TO_ASCII
```
Comment (TheOneRing) [inline src/libsync/CMakeLists.txt:81]: private

File: test/csync/CMakeLists.txt [hunk 9]
```diff
@@ -1,15 +1,16 @@
...
-  ${CMOCKA_INCLUDE_DIR}
-)
 
-include_directories(${CHECK_INCLUDE_DIRS})
 # create test library
 add_library(torture STATIC torture.c cmdline.c)
 target_link_libraries(torture ${CMOCKA_LIBRARIES})
 set_target_properties(torture PROPERTIES C_STANDARD 99 AUTOMOC OFF)
+target_include_directories(torture
+    PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}
```
Comment (TheOneRing) [inline test/csync/CMakeLists.txt:8]: https://cmake.org/cmake/help/latest/variable/CMAKE_INCLUDE_CURRENT_DIR.html

And the src dir should already be included by linking to csync.

File: src/csync/CMakeLists.txt (hunk 5 above)
Comment (TheOneRing) [reply src/csync/CMakeLists.txt:37]: One PUBLIC should be enough

**Review by TheOneRing — CHANGES_REQUESTED** [review]

File: shell_integration/dolphin/CMakeLists.txt (hunk 1 above)
Comment (fmoc) [reply shell_integration/dolphin/CMakeLists.txt:43]: Let me check whether it works without this dir.

I think in Qt projects, `CMAKE_INCLUDE_CURRENT_DIR` should be set to `On` anyway.

File: src/csync/CMakeLists.txt (hunk 5 above)
Comment (fmoc) [reply src/csync/CMakeLists.txt:37]: Indeed, it's a code style thing. I'd rather move the other `PUBLIC` public flags to new lines as well.

File: src/csync/CMakeLists.txt (hunk 4 above)
Comment (fmoc) [reply src/csync/CMakeLists.txt:45]: Which of those? They used to be used in all binaries.

File: src/csync/CMakeLists.txt (hunk 7 above)
Comment (fmoc) [reply src/csync/CMakeLists.txt:61]: Should we remove it? We could use a `message(FATAL_ERROR ...`).
(In fact, not using `message(STATUS ...)` is a problem already.)

File: src/libsync/CMakeLists.txt (hunk 8 above)
Comment (fmoc) [reply src/libsync/CMakeLists.txt:81]: All of them? Only this one?

File: shell_integration/dolphin/CMakeLists.txt (hunk 2 above)
Comment (fmoc) [reply shell_integration/dolphin/CMakeLists.txt:44]: I could only make them private when repeating this code for the other two targets.

File: shell_integration/dolphin/CMakeLists.txt (hunk 1 above)
Comment (TheOneRing) [reply shell_integration/dolphin/CMakeLists.txt:43]: It is through KDE settings

File: shell_integration/dolphin/CMakeLists.txt (hunk 2 above)
Comment (TheOneRing) [reply shell_integration/dolphin/CMakeLists.txt:44]: Please do so, setting compile flags is not duplicating code and such build flags should never be propagated to all dependent targets.

File: src/csync/CMakeLists.txt (hunk 4 above)
Comment (TheOneRing) [reply src/csync/CMakeLists.txt:45]: Some yes, but some where only set on a directory level.
QT_NO_CAST_TO_ASCII will cause all QString("foo") calls to fail which are still used in the gui code.

File: src/libsync/CMakeLists.txt (hunk 8 above)
Comment (fmoc) [reply src/libsync/CMakeLists.txt:81]: We have to remove the first one altogether, the rest should be private, in order to make the build work.

File: src/csync/CMakeLists.txt (hunk 7 above)
Comment (TheOneRing) [reply src/csync/CMakeLists.txt:61]: Would require removing the ifdef from all the code, keep it for now pls.
I was just mentioning my resent...

File: src/libsync/CMakeLists.txt (hunk 8 above)
Comment (TheOneRing) [reply src/libsync/CMakeLists.txt:81]: All

**Review by TheOneRing — APPROVED** [review]

**Comment by individual-it** [comment]
@TheOneRing @fmoc 
are these changes intended?
**before this PR**
grafik

**after this PR**
grafik

**Comment by TheOneRing** [comment]
Its not, looks like the application name isn't set correctly.
@fmoc pls fix

**Pull request merged**